bpf/process/types/probe_read_kernel_or_user.h

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+/* Copyright Authors of Cilium */
+
+#ifndef __PROBE_READ_KERNEL_OR_USER_H__
+#define __PROBE_READ_KERNEL_OR_USER_H__
+
+#include "bpf_core_read.h"
+
+#define bpf_probe_read_kernel probe_read_kernel
+#define bpf_probe_read_user   probe_read_user
+
+#ifdef __LARGE_BPF_PROG
+static inline __attribute__((always_inline)) int
+probe_read_kernel_or_user(void *dst, uint32_t size, const void *src, bool userspace)
+{
+	if (userspace)
+		return probe_read_user(dst, size, src);
+	return probe_read_kernel(dst, size, src);
+}
+
+static inline __attribute__((always_inline)) int
+probe_read_kernel_or_user_masked(void *dst, uint32_t size, uint32_t size_mask, const void *src, bool userspace)
+{
+	if (userspace) {
+		asm volatile("%[size] &= %1;\n"
+			     : [size] "+r"(size)
+			     : "i"(size_mask));
+		return probe_read_user(dst, size, src);
+	}
+	asm volatile("%[size] &= %1;\n"
+		     : [size] "+r"(size)
+		     : "i"(size_mask));
```

Comment:
Is there a reason for doing the assembly part twice? It's just a single instruction, but wouldn't it be better to do it once and then do the branch?

E.g.,
```
	asm volatile("%[size] &= %1;\n"
			     : [size] "+r"(size)
			     : "i"(size_mask));
	return userspace ? probe_read_user(dst, size, src) : probe_read_kernel(dst, size, src);
```

Reply:
Yep, because otherwise the registers spill during the branch and then the masking is forgotten (on older kernels) when the registers are restored from the stack.